### PR TITLE
Add documentation to invoke a service idempotently

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Documentation for https://github.com/restatedev/restate/issues/784

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
To be merged when releasing 0.4

File: docs/services/invocation.md [hunk 1]
```diff
@@ -67,6 +67,21 @@ grpcurl -plaintext <your-restate-runtime-host-port> describe
...
+```shell
+curl -X POST http://<restate-runtime-host-port>/org.example.Greeter/Greet -H 'idempotency-key: <idempotency-key>' -H 'content-type: application/json' -d '{"name": "Pete"}'
+```
+
+After the invocation completes, Restate persists the response for a retention period of 30 minutes.
+If you re-invoke the service with the same idempotency key within 30 minutes, Restate sends back the same response and doesn't re-execute the request to the service.
+
+You can tune the retention period by setting the header `idempotency-retention-period: <seconds>`
+
+The `Idempotency-Key` header works with both HTTP and gRPC/gRPC-web.
```
Comment (tillrohrmann) [inline docs/services/invocation.md:83]: Really cool feature @slinkydeveloper! Also that it works transparently for HTTP and gRPC is great.

**Review by tillrohrmann — APPROVED** [review]
LGTM. +1 for merging once we create the release.

File: docs/services/invocation.md (hunk 1 above)
Comment (AhmedSoliman) [reply docs/services/invocation.md:83]: 30 minutes default might be too short. Is this configurable somewhere? Stripe for instance has 24 hours of idempotency key validity, there were even discussions to extend this to 30 days :)

Comment (slinkydeveloper) [reply docs/services/invocation.md:83]: The default is hardcoded, we can make it configurable though. You can also set it up per request with the aforementioned header. Changing to 24 hours sounds good to me.

Comment (AhmedSoliman) [reply docs/services/invocation.md:83]: Maybe it's a good time to discuss where such configuration should live. Idempotency key is part of the service public contract, maybe naively, I expect it to be part of the "Service Configuration" if that exists.

Comment (slinkydeveloper) [reply docs/services/invocation.md:83]: > I expect it to be part of the "Service Configuration" if that exists.

Nothing like that currently exists. The only bit you can configure about a service is its instance type, in the service contract. 

>  Idempotency key is part of the service public contract

Maybe? Let me elaborate a bit my train of thoughts:

* I think there's a bit of misunderstanding of the scope of this feature: this currently works only for invocations **from outside world**, that is from the ingress. It effectively allows you to execute calls to services exactly once, in a similar fashion to how you have invocations between restate services: if you invoke from service A the method B::Do, that invocation will be executed once. With this feature, if you invoke from HTTP client X with idempotency key A the method B::Do, that invocation will be executed once (given the timeout didn't expired). 
* Nothing stops us from supporting this feature for invocations between restate services (needs a bit of effort, but not too much), but the question I ask myself is: Should we do that?
* Now there can be cases where both service A and C and HTTP client X with idempotency key A invoke B::Do. In the current state of things, it's up to B::Do to make sure it won't re-execute its business logic. But isn't this a "different level" of idempotency, more at a business logic level and less on a physical request level? Would this needs some special business logic from the user to handle this? For example, what response should you send in this case? Something like "Not Modified"? Would in fact this situation still apply some side effects (e.g. update some "modification time"), but not re-apply all the original side effects?

**Pull request merged**